hearth/lib/hearth/middleware/retry.rb

```diff
@@ -2,114 +2,93 @@
+
+        raise 'Unable to make request, too many failures' if token.nil?
+
+        # TODO: determine if loop or nested function call is better here
+        # If the latter, no outer variable is needed.
+        output = nil
+        loop do
+          output = @app.call(input, context)
+
+          if (error = output.error)
+            error_inspector = @error_inspector_class.new(
+              error, context.response
+            )
+            error_info = error_info(error, error_inspector)
+            token = @retry_strategy.refresh_retry_token(token, error_info)
+
+            Kernel.sleep(token.retry_delay) if token
+          else
+            @retry_strategy.record_success(token)
+          end
+          break unless token && output.error
+
+          reset_request(context, output)
+        end
+        output
       end
-      # rubocop:enable Metrics/AbcSize
 
       private
 
-      def acquire_token
-        return if @retry_mode == 'standard'
-
-        # either fail fast or block until a token becomes available
-        # must be configurable
-        # need a maximum rate at which we can send requests (max_send_rate)
-        # is unset until a throttle is seen
-        @client_rate_limiter.token_bucket_acquire(
-          1, wait_to_fill: @adaptive_retry_wait_to_fill
-        )
+      def error_info(error, error_inspector)
+        {
+          error_type: error_type(error, error_inspector),
+          retryable: modeled_retryable?(error)
+        }.merge(error_inspector.hints)
       end
 
-      # max_send_rate is updated if on adaptive mode and based on response
-      # retry quota is updated if the request is successful (both modes)
-      def request_bookkeeping(output, error_inspector)
-        @retry_quota.release(@capacity_amount) unless output.error
-
-        return unless @retry_mode == 'adaptive'
-
-        @client_rate_limiter.update_sending_rate(
-          error_inspector.error_type == 'Throttling'
-        )
+      def error_type(error, error_inspector)
+        if error_inspector.transient?
+          'Transient'
+        elsif modeled_throttling?(error) || error_inspector.throttling?
+          'Throttling'
+        elsif error_inspector.server?
+          'ServerError'
+        elsif error_inspector.client?
+          'ClientError'
+        else
+          'UnknownError'
+        end
       end
 
-      def retryable?(context, output, error_inspector)
-        return false unless output.error
+      def modeled_retryable?(error)
+        error.is_a?(Hearth::ApiError) && error.retryable?
+      end
 
-        error_inspector.retryable? &&
-          context.response.body.respond_to?(:truncate)
+      def modeled_throttling?(error)
+        modeled_retryable?(error) && error.throttling?
       end
 
-      def retry_request(input, context, output)
-        @retries += 1
+      # TODO: find a way to dup req/resp, to undo signing, rather than resetting?
```

Comment:
Not sure what's preferred here. We may want to keep a copy of the request before signing, and use that for retries. The object IDs might be different though. Right now we just reset everything ourselves.